### ISSUE TITLE
qpdf: INTERNAL ERROR: QPDFWriter::generateID has no data for deterministic ID.

### ISSUE THREAD

**Issue by fekir** [qpdf/qpdf#1235]
> qpdf: INTERNAL ERROR: QPDFWriter::generateID has no data for deterministic ID.  This may happen if deterministic ID and file encryption are requested together.

This error happens when adding the flag `--deterministic-id`, and qpdf generates an empty file.

The document is pretty big (over 9MB), and not public, but is not encrypted.

Since any constant string could be a valid id, why not use a fallback value instead of aborting?
(eventually leave the warning)

Got the error with `qpdf version 11.9.1` packaged in debian

**Comment by fekir** [comment]
Sorry, the document is encrypted (I need to check what it actually means as I do not need to enter a password to see the content).

Nevertheless, I see little reason for `--deterministic-id` to fail in this situation.

**Comment by fekir** [comment]
Current workaround is to use `--decrypt` together with `--deterministic-id`.

I'm not interested in the encryption, and `--decrypt` does not lead to an error if the file is not encrypted.

Nevertheless, it would be better if `--deterministic-id` would work without `--decrypt`

**Comment by m-holger** [comment]
I agree that qpdf should not throw an internal error in this situation, nor should an empty file be produced. Thanks for reporting this.

> Since any constant string could be a valid id, why not use a fallback value instead of aborting?

Producing a constant string would not be conformant with the PDF specification and therefore is not an option. qpdf has the option --static-id which essentially does what you are suggesting, but this option **must only be used to create files for testing purposes** and never to produce real life PDF files.

> Sorry, the document is encrypted (I need to check what it actually means as I do not need to enter a password to see the content).

A brief introduction can be found at https://qpdf.readthedocs.io/en/latest/encryption.html

> Nevertheless, I see little reason for --deterministic-id to fail in this situation.

The difficulty arises from the fact that ids have to be unique. To achieves this, deterministic-id has to fingerprint the file in order to ensure that ids for different files don't clash, and therefore can only be generated after all the significant parts of the output file have been written. On the other hand, the id is required to generate encryption keys and therefore (for encrypted files)  is required before writing can start. This leaves us with a chicken and egg situation.

**Comment by fekir** [comment]
An example document can be found here

https://uwaterloo.ca/onbase/sites/ca.onbase/files/uploads/files/samplesecured_256bitaes_pdf.pdf

> deterministic-id has to fingerprint the file in order to ensure that ids for different files don't clash

Just thinking out loud, an alternative could be to fingerprint the input parameters (and thus also the referenced files) of qpdf, instead of the output.

> On the other hand, the id is required to generate encryption keys and therefore (for encrypted files) is required before writing can start.

Again, thinking out loud: create a file without encryption, then encrypt it.
Not nice as it means to create two files, but also not necessary if the id is based on the input.

**Comment by m-holger** [comment]
Fix is discussed at https://github.com/qpdf/qpdf-dev/discussions/9

**Comment by m-holger** [comment]
@fekir , out of interest, what is your aim of using --deterministic-id?

>  an alternative could be to fingerprint the input parameters (and thus also the referenced files) of qpdf

This might work, but the effort would be disproportionate. The only way I can see this happening is if you submit a PR with your implementation.

> create a file without encryption, then encrypt it

I don't think this would work for a number of reasons, including that it would not guaranteed unique ids, which is a show-stopper. Also,  an implementation limitation of qpdf's writer is that it does not support writing a QPDF object more than once.

**Comment by fekir** [comment]
As hinted at https://github.com/qpdf/qpdf/issues/1165 - reproducible builds (the PDF is an artifact, and I want it to be reproducible)

In this particular case, I wanted to reduce the size of some files, I think I used something similar to 

~~~
qpdf --recompress-flate --compress-streams=y --object-streams=generate --remove-unreferenced-resources=yes --coalesce-contents
~~~

Those files are synchronized between different devices (and some might end in some archives) I want to know quickly if two files are equal or not, before synchronizing them.


If the compression algorithms will ever be improved (for example though https://github.com/qpdf/qpdf/issues/1164), and I'm going to try to reduce the size of the PDF files again, I want to see in a diff only the list of files that changed, but not because the internal id changed and the rest of the file not.

Sure, I could use a smart diff; verify if the content is different or not, but it would be too slow after a certain amount of files, and if I need to inspect files archived somewhere else.

Personally, I have no interest in encrypted files; I was not even aware that some of my files where encrypted, thus using `--decrypt` was not an issue; on the contrary.


To sum it up: I do not believe most people will be affected by this, and there are multiple workarounds.

> it would not guaranteed unique ids

Well, no algorithm will guarantee that.
Especially as anyone can edit manually a document to match another, it should thus be a "best-effort".

Out of curiosity; is there any system that relies on such id to be unique?

**Comment by jberkenbilt** [comment]
> Fix is discussed at [qpdf/qpdf-dev#9](https://github.com/qpdf/qpdf-dev/discussions/9)

I think we should discuss the fix here, not on a separate issue in qpdf-dev.

**Comment by m-holger** [comment]
> Out of curiosity; is there any system that relies on such id to be unique?

I don't know. The beauty of standards is that I do not need to know what everybody else is doing. All conformant tools will play together nicely. If we pick and choose which parts of the standard we like, some of the tools will play together some of the time, to some degree.

**Comment by jberkenbilt** [comment]
The deterministic ID feature was implemented as a specific request from Google, who, at the time, used qpdf to linearize files when you exported them from Google Docs. qpdf won an open source award from Google for this enhancement. A few years ago, they stopped doing this, so Google Docs exported PDF files no longer pass through qpdf. The reason they wanted deterministic ID was not to use the ID for any purpose itself but because having an identical PDF generated from two exports of an unchanged Google Doc made caching work better.

I am not personally aware of any system that relies on ID being unique. Relying on this would be unwise as it would rest upon all the PDFs you see having been generated by people who didn't make a mistake about ID generation, and there are a vast number of incorrect PDFs out there. Also, ID is optional per the spec. And, in theory, you can have has collisions, though the probability is low without explicit attempts to exploit broken MD5.

Anyway, none of this is to say that IDs shouldn't be unique. The intention is that they are probably unique, and as m-holger says, we have no way to know or no need to know how people use or misuse these. If someone tried to rely on ID being unique and had a relatively consistent collection of relatively valid PDFs, as long as they tolerated the occasional collision, they would probably do just fine.

**Comment by m-holger** [comment]
Or, to quote @jberkenbilt  

> In general, one should adhere strictly to a specification when writing but be liberal in reading. This way, the product of our software will be accepted by the widest range of other programs, and we will accept the widest range of input files.
